src/InodeCache.cpp

```diff
@@ -140,6 +144,35 @@ fd_is_on_known_to_work_file_system(int fd)
   return known_to_work;
 }
 
+bool
+spin_lock(std::atomic<pid_t>& owner_pid, const pid_t self_pid)
+{
+  for (;;) {
+    for (int i = 0; i < 10000; ++i) {
+      pid_t pid = 0;
+      if (owner_pid.load(std::memory_order_relaxed) == 0
+          && owner_pid.compare_exchange_weak(
+            pid, self_pid, std::memory_order_acquire)) {
+        return false;
+      }
+      sched_yield();
+    }
+    pid_t lock_owner = owner_pid.load(std::memory_order_relaxed);
+    if (lock_owner != 0 && getpgid(lock_owner) == -1 && errno == ESRCH) {
```

Comment:
Why `getpgid` here when `getpid` is used to set `m_self_pid`?

Reply:
Because you can't use getpid to to check if a different pid is still alive while getpgid will error if not?

Reply:
I see, it's a variation of `kill(pid, 0)` that I haven't seen before.

Reply:
Process may have not enough rights to send a kill, however getpgid is always allowed.

Reply:
For the record note that also getpgid may fail with EPERM by spec:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/getpgid.html

But regardless of using getpgid or kill, EPERM is most likely indicating the process is still around and then it should be the right decision to not remove the lock since that assumption can't cause false positives even if wrong.